app/controllers/payment_controller.rb

```diff
@@ -1,43 +1,84 @@
 # frozen_string_literal: true
+
 class PaymentController < ApplicationController
   skip_before_action :verify_authenticity_token
 
   def transaction
-    builder = if recurring?
-                client::Subscription.make_subscription(payment_options)
-              else
-                client::Transaction.make_transaction(payment_options)
-              end
-
     if builder.success?
-      write_member_cookie(builder.action.member_id) unless builder.action.blank?
-      id = recurring? ? { subscription_id: builder.subscription_id } : { transaction_id: builder.transaction_id }
-
-      respond_to do |format|
-        format.html { redirect_to follow_up_page_path(page) }
-        format.json { render json: { success: true }.merge(id) }
-      end
+      process_and_render_success
     else
-      @errors = client::ErrorProcessing.new(builder.error_container, locale: locale).process
-      @page = page
-      respond_to do |format|
-        format.html { render 'payment/donation_errors', layout: 'member_facing' }
-        format.json { render json: { success: false, errors: @errors }, status: 422 }
-      end
+      render_errors
     end
   end
 
+  def process_and_render_success
+    write_member_cookie(builder.action.member_id) unless builder.action.blank?
+    id = recurring? ? { subscription_id: builder.subscription_id } : { transaction_id: builder.transaction_id }
+
+    if store_in_vault?
+      result = BraintreeServices::PaymentResult.new(builder.result)
+
+      existing_payment_methods = (cookies.signed[:payment_methods] || '').split(',')
+      (existing_payment_methods << result.payment_method_token).uniq
```

Comment:
confused by this line because the call to `uniq` is not in place so I'm not sure where the result is being stored